src/Facets/GenericSwapFacet.sol

```diff
@@ -3,42 +3,328 @@ pragma solidity 0.8.17;
+import { console2 } from "forge-std/console2.sol";
 
-/// @title Generic Swap Facet
+/// @title GenericSwapFacet
 /// @author LI.FI (https://li.fi)
 /// @notice Provides functionality for swapping through ANY APPROVED DEX
-/// @dev Uses calldata to execute APPROVED arbitrary methods on DEXs
-/// @custom:version 1.0.0
+/// @dev Can only execute calldata for APPROVED function selectors
+/// @custom:version 2.0.0
 contract GenericSwapFacet is ILiFi, ReentrancyGuard, SwapperV2, Validatable {
+    using SafeERC20 for IERC20;
+
+    error InsufficientSwapOutput();
+
+    event AssetSwapped(
+        bytes32 transactionId,
+        address dex,
+        address fromAssetId,
+        address toAssetId,
+        uint256 fromAmount,
+        uint256 toAmount,
+        uint256 timestamp
+    );
+
     /// External Methods ///
 
-    /// @notice Performs multiple swaps in one transaction
+    /// @notice Performs a single swap from an ERC20 token to another ERC20 token
     /// @param _transactionId the transaction id associated with the operation
     /// @param _integrator the name of the integrator
     /// @param _referrer the address of the referrer
     /// @param _receiver the address to receive the swapped tokens into (also excess tokens)
-    /// @param _minAmount the minimum amount of the final asset to receive
+    /// @param _minAmountOut the minimum amount of the final asset to receive
+    /// @param _swapData an object containing swap related data to perform swaps before bridging
+    function swapTokensSingleERC20ToERC20(
+        bytes32 _transactionId,
+        string calldata _integrator,
+        string calldata _referrer,
+        address payable _receiver,
+        uint256 _minAmountOut,
+        LibSwap.SwapData calldata _swapData
+    ) external payable nonReentrant returns (uint256 amountOut) {
+        // deposit funds
+        IERC20(_swapData.sendingAssetId).safeTransferFrom(
+            msg.sender,
+            address(this),
+            _swapData.fromAmount
+        );
+
+        // ensure that contract (callTo) and function selector are whitelisted
+        if (
+            !(LibAllowList.contractIsAllowed(_swapData.callTo) &&
+                LibAllowList.selectorIsAllowed(bytes4(_swapData.callData[:4])))
+        ) revert ContractCallNotAllowed();
+
+        // get initial token balance
+        uint256 initialBalance = IERC20(_swapData.receivingAssetId).balanceOf(
+            address(this)
+        );
+
+        // check if the current allowance is sufficient
+        if (
+            IERC20(_swapData.sendingAssetId).allowance(
+                address(this),
+                _swapData.approveTo
+            ) < _swapData.fromAmount
+        ) {
+            // allowance insufficient - register max approval
+            SafeERC20.safeApprove(
+                IERC20(_swapData.sendingAssetId),
+                _swapData.approveTo,
+                0
+            );
```

Comment:
Ideally this is never needed as we have either no approval or max approval. Does it make sense to handle that case explicitly
```
if (approval === 0)
  set max
else if (approval < fromAmount)
  reset + set max
```

Reply:
yes I like that. Thank you.